### PR TITLE
Support !=, like filter expressions in OAL

### PR DESCRIPTION
**Pull request by kezhenxu94**
And slightly refactor the matchers

resolves https://github.com/apache/skywalking/issues/5234

### CODE REVIEW

File: oap-server/oal-rt/src/main/java/org/apache/skywalking/oal/rt/OALRuntime.java [hunk 1]
```diff
@@ -150,6 +151,12 @@ public void start(ClassLoader currentClassLoader) throws ModuleStartException, O
             throw new ModuleStartException("load metrics functions error.", e);
         }
 
+        try {
+            FilterMatchers.INSTANCE.init();
```
Comment (arugal) [inline oap-server/oal-rt/src/main/java/org/apache/skywalking/oal/rt/OALRuntime.java:155]: The `start` method is called more than once

Comment (kezhenxu94) [reply oap-server/oal-rt/src/main/java/org/apache/skywalking/oal/rt/OALRuntime.java]: No side effect even initializing multiple times, fixed anyway

File: oap-server/oal-rt/src/main/java/org/apache/skywalking/oal/rt/parser/MetricsHolder.java [hunk 2]
```diff
@@ -26,10 +26,19 @@
...
+    private static final Map<String, Class<? extends Metrics>> REGISTER = new HashMap<>();
 
-    public static void init() throws IOException {
+    static {
+        try {
+            init();
+        } catch (IOException e) {
+            throw new RuntimeException("load metrics functions error.", e);
+        }
+    }
```
Comment (wu-sheng) [inline oap-server/oal-rt/src/main/java/org/apache/skywalking/oal/rt/parser/MetricsHolder.java:39]: Please make this as lazy initialization in the `#find` method. The static block is not recommended.

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/metrics/expression/LikeMatch.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+ * limitations under the License.
+ *
+ */
+
+package org.apache.skywalking.oap.server.core.analysis.metrics.expression;
+
+import org.apache.skywalking.oap.server.core.analysis.metrics.annotation.FilterMatcher;
+
+@FilterMatcher
+public class LikeMatch {
```
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/metrics/expression/LikeMatch.java:24]: `like` should support `%xxx%`, right?

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/metrics/annotation/FilterMatcher.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+     * the default value is the name of the class annotated with {@link FilterMatcher}, with the first letter being
+     * lowercase.
+     */
+    String[] value() default {};
+
+    /**
+     * @return the type of the filter operand, which decides the getter of the operands, i.e. getter of {@code Boolean}
+     * and {@code boolean} type is {@code isArg}, otherwise {@code getArg}.
+     */
+    Class<?> type() default Object.class;
```
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/metrics/annotation/FilterMatcher.java:45]: According to the usage and the comments, why just name this as `isBoolean`? Or, I prefer to create a new `BooleanValueFilterMatcher` annotation?

File: oap-server/oal-rt/src/main/java/org/apache/skywalking/oal/rt/parser/DeepAnalysis.java [hunk 5]
```diff
@@ -45,45 +45,22 @@ public AnalysisResult analysis(AnalysisResult result) {
...
-                    filterExpression.setExpressionObject("LessEqualMatch");
-                    filterExpression.setLeft("source." + ClassMethodUtil.toGetMethod(expression.getAttribute()) + "()");
-                    filterExpression.setRight(expression.getValue());
-                    result.addFilterExpressions(filterExpression);
-                } else {
-                    throw new IllegalArgumentException("filter expression [" + expression.getExpressionType() + "] not found");
-                }
+                final FilterMatchers.MatcherInfo matcherClass = FilterMatchers.INSTANCE.find(expression.getExpressionType());
+
+                final String getter = matcherClass.getType() == boolean.class || matcherClass.getType() == Boolean.class
```
Comment (wu-sheng) [inline oap-server/oal-rt/src/main/java/org/apache/skywalking/oal/rt/parser/DeepAnalysis.java:50]: Replacing `getType()` to `isBoolean` seems more clear.

File: oap-server/oal-rt/src/main/java/org/apache/skywalking/oal/rt/parser/FilterMatchers.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.skywalking.oap.server.core.analysis.metrics.annotation.BooleanValueFilterMatcher;
+import org.apache.skywalking.oap.server.core.analysis.metrics.annotation.FilterMatcher;
+
+@SuppressWarnings("UnstableApiUsage")
```
Comment (wu-sheng) [inline oap-server/oal-rt/src/main/java/org/apache/skywalking/oal/rt/parser/FilterMatchers.java:32]: Which unstable API?

File: oap-server/oal-rt/src/main/java/org/apache/skywalking/oal/rt/parser/MetricsHolder.java [hunk 7]
```diff
@@ -45,13 +48,16 @@ public static void init() throws IOException {
         }
     }
 
-    public static Class<? extends Metrics> find(
-        String functionName) {
-        String func = functionName;
-        Class<? extends Metrics> metricsClass = REGISTER.get(
-            func);
+    @SneakyThrows
```
Comment (wu-sheng) [inline oap-server/oal-rt/src/main/java/org/apache/skywalking/oal/rt/parser/MetricsHolder.java:51]: AFAIK, IllegalArgumentException is a RuntimeException. Why need this?

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/metrics/annotation/FilterMatcher.java (hunk 4 above)
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/metrics/annotation/FilterMatcher.java:45]: Do you forget to delete this?

File: oap-server/oal-rt/src/main/java/org/apache/skywalking/oal/rt/parser/FilterMatchers.java (hunk 6 above)
Comment (kezhenxu94) [reply oap-server/oal-rt/src/main/java/org/apache/skywalking/oal/rt/parser/FilterMatchers.java:32]: `com.google.common.reflect.ClassPath`

File: oap-server/oal-rt/src/main/java/org/apache/skywalking/oal/rt/parser/MetricsHolder.java (hunk 7 above)
Comment (kezhenxu94) [reply oap-server/oal-rt/src/main/java/org/apache/skywalking/oal/rt/parser/MetricsHolder.java:51]: `init` throws `IOException`, which is not runtime exception

**Pull request merged**